File: 2024/with_python/src/day2.py
def part1(content: list[str]) -> int:
    return sum(1 for report in content if is_sorted(report) and has_valid_gaps(report))


def part2(content: list[str]) -> int:
    return sum(
        1 for report in content if is_sorted_and_valid_gaps_with_exception(report)
    )


def is_sorted(report: str) -> bool:
    nums = [int(x) for x in report.split()]
    return nums == sorted(nums) or nums == sorted(nums, reverse=True)


def has_valid_gaps(report: str) -> bool:
    nums = [int(x) for x in report.split()]
    return all(1 <= abs(nums[i] - nums[i - 1]) <= 3 for i in range(1, len(nums)))


def is_sorted_and_valid_gaps_with_exception(report: str) -> bool:
    if is_sorted(report) and has_valid_gaps(report):
        return True

    nums = report.split()
    return any(
        is_sorted(" ".join(nums[:i] + nums[i + 1 :]))
        and has_valid_gaps(" ".join(nums[:i] + nums[i + 1 :]))
        for i in range(len(nums))
    )

File: 2024/with_rust/src/day2.rs
use itertools::Itertools;

pub fn part1(content: Vec<String>) -> usize {
    content
        .iter()
        .filter(|report| is_sorted(report) && has_valid_gaps(report))
        .count()
}

pub fn part2(content: Vec<String>) -> usize {
    content
        .iter()
        .filter(|report| is_sorted_and_valid_gaps_with_one_exception(report))
        .count()
}

fn is_sorted(report: &str) -> bool {
    let numbers: Vec<i32> = from_str_to_numbers(report);

    numbers.is_sorted() || numbers.is_sorted_by(|a, b| b < a)
}

fn has_valid_gaps(report: &str) -> bool {
    let numbers: Vec<i32> = from_str_to_numbers(report);

    numbers.windows(2).all(|pair| {
        let delta = (pair[1] - pair[0]).abs();
        delta >= 1 && delta <= 3
    })
}

fn from_str_to_numbers(report: &str) -> Vec<i32> {
    report.split_whitespace().flat_map(str::parse).collect()
}

fn is_sorted_and_valid_gaps_with_one_exception(report: &str) -> bool {
    if is_sorted(report) && has_valid_gaps(report) {
        return true;
    }
    let numbers: Vec<i32> = from_str_to_numbers(report);
    for i in 0..numbers.len() {
        let mut new_numbers = numbers.clone();
        new_numbers.remove(i);

        if is_sorted(&new_numbers.iter().join(" ")) && has_valid_gaps(&new_numbers.iter().join(" "))
        {
            return true;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::reader;

    #[test]
    fn part1solution() {
        assert_eq!(part1(reader::read_input("day2sample")), 2);
    }

    #[test]
    fn part2solution() {
        assert_eq!(part2(reader::read_input("day2sample")), 4);
    }
}
